Testnet/Airdrop/Airdrop.Tests/AirdropTests.cs

```diff
@@ -0,0 +1,785 @@
+            Assert.False(result);
+        }
+
+        [Fact]
+        public void Register_Fail_MoreRegistrantsThanSupply()
+        {
+            this.mockContractState.Setup(m => m.Message).Returns(new Message(this.airdropContract, this.registrant, 0));
+            this.mockContractState.Setup(b => b.Block.Number).Returns(1);
+            this.mockPersistentState.Setup(e => e.GetUInt64("EndBlock")).Returns(this.endBlock);
+            this.mockPersistentState.Setup(e => e.GetUInt64("NumberOfRegistrants")).Returns(10);
+            this.mockPersistentState.Setup(e => e.GetUInt64("TotalSupply")).Returns(1);
+
+            // Initialize the smart contract set constructor props
+            var airdrop = new Airdrop(this.mockContractState.Object, this.tokenContract, 1, this.endBlock);
+
+            var result = airdrop.Register();
+            Assert.False(result);
+
+            this.mockPersistentState.Verify(s => s.GetStruct<Status>($"Status:{this.registrant}"));
+            this.mockPersistentState.Verify(s => s.GetBool($"RegistrationIsClosed"));
+            this.mockPersistentState.Verify(s => s.GetUInt64($"TotalSupply"));
+            this.mockPersistentState.Verify(s => s.GetUInt64($"NumberOfRegistrants"));
+
+            var registrationIsClosed = airdrop.RegistrationIsClosed;
+            Assert.False(registrationIsClosed);
+
+            var accountStatus = airdrop.GetAccountStatus(this.registrant);
+            Assert.Equal(Status.NOT_ENROLLED, accountStatus);
+
+            var _totalSupply = airdrop.TotalSupply;
+            Assert.Equal((ulong)1, _totalSupply);
+
+            var numRegistrants = airdrop.NumberOfRegistrants;
+            Assert.Equal((ulong)10, numRegistrants);
+        }
+
+        [Fact]
+        public void Register_Success_NumberOfRegistrantsIncrement()
+        {
+            this.mockContractState.Setup(m => m.Message).Returns(new Message(this.airdropContract, this.registrant, 0));
+            this.mockContractState.Setup(b => b.Block.Number).Returns(999_999);
+            this.mockPersistentState.Setup(e => e.GetUInt64("NumberOfRegistrants")).Returns(this.index);
+            this.mockPersistentState.Setup(e => e.GetUInt64("EndBlock")).Returns(this.endBlock);
+            this.mockPersistentState.Setup(e => e.GetUInt64("TotalSupply")).Returns(this.totalSupply);
+
+            // Initialize the smart contract set constructor props
+            var airdrop = new Airdrop(this.mockContractState.Object, this.tokenContract, this.totalSupply, this.endBlock);
+
+            var result = airdrop.Register();
+            this.mockPersistentState.Verify(s => s.SetUInt64("NumberOfRegistrants", airdrop.NumberOfRegistrants + 1));
+
+            for (uint i = 1; i < 5; i++)
+            {
+                this.mockPersistentState.Setup(s => s.GetUInt64("NumberOfRegistrants")).Returns(i);
+                Assert.Equal(i, airdrop.NumberOfRegistrants);
+            }
+        }
+
+        //[Fact]

```

Comment:
```suggestion
          [Fact]
        public void Register_Fail_NumberOfRegistrantsNotIncremented()
        {
            ulong totalNumberOfRegistrants = 0;
            this.mockContractState.Setup(m => m.Message).Returns(new Message(this.airdropContract, this.airdropContractOwner, 0));
            this.mockContractState.Setup(b => b.Block.Number).Returns(999_999);
            this.mockPersistentState.Setup(e => e.GetUInt64("NumberOfRegistrants")).Returns(() => { return totalNumberOfRegistrants; });
            this.mockPersistentState.Setup(e => e.SetUInt64("NumberOfRegistrants", It.IsAny<ulong>())).Callback<string, ulong>((key, value) => { totalNumberOfRegistrants = value; });
            this.mockPersistentState.Setup(e => e.GetUInt64("EndBlock")).Returns(this.endBlock);
            this.mockPersistentState.Setup(e => e.GetUInt64("TotalSupply")).Returns(this.totalSupply);

            Dictionary<string, Status> registrations = new Dictionary<string, Status>();
            this.mockPersistentState.Setup(e => e.GetStruct<Status>($"Status:{this.registrant}"))
                .Returns<string>((key) =>
                {
                    if (registrations.ContainsKey(key))
                    {
                        return registrations[key];
                    }

                    return Status.NOT_ENROLLED;
                });
            this.mockPersistentState.Setup(e => e.SetStruct($"Status:{this.registrant}", It.IsAny<Status>()))
                .Callback<string, Status>((key, value) =>
                {
                    if (registrations.ContainsKey(key))
                    {
                        registrations[key] = value;
                    }
                    else
                    {
                        registrations.Add(key, value);
                    }
                });

            // Initialize the smart contract set constructor props
            var airdrop = new Airdrop(this.mockContractState.Object, this.tokenContract, this.totalSupply, this.endBlock);

            Assert.Equal((ulong)0, totalNumberOfRegistrants);
            this.mockContractState.Setup(m => m.Message).Returns(new Message(this.airdropContract, this.registrant, 0));

            // call first time with registrant that is not registered yet.
            var result = airdrop.Register();

            Assert.True(result);
            this.mockPersistentState.Verify(s => s.SetUInt64("NumberOfRegistrants", It.IsAny<ulong>()), Times.Once);
            Assert.Equal((ulong)1, totalNumberOfRegistrants);

            // call second time with same registrant that is already registered.
            result = airdrop.Register();
            Assert.False(result);
            Assert.Equal((ulong)1, totalNumberOfRegistrants);

            // verify it's not called more times than before. It should still be only once.
            this.mockPersistentState.Verify(s => s.SetUInt64("NumberOfRegistrants", It.IsAny<ulong>()), Times.Once);
        }
```

I suggest we start setting up these tests with a mocked storage behind them, see how I've done it. That will give more realistic test results as they now depend on a lot of 'static' mocking.